internal/pkg/gateway/api.go

```diff
@@ -86,24 +91,90 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
+	var responses []*peer.ProposalResponse
+	if len(request.EndorsingOrganizations) > 0 {
+		// The client is specifying the endorsing orgs and taking responsibility for ensuring it meets the signature policy
+		endorsers, err = gs.registry.endorsersForOrgs(channel, chaincodeID, request.EndorsingOrganizations)
+		if err != nil {
+			return nil, status.Errorf(codes.Unavailable, "%s", err)
+		}
+	} else {
+		// The client is delegating choice of endorsers to the gateway.
+
+		// 1. Choose an endorser from the gateway's organization
+		var firstEndorser *endorser
+		es, ok := gs.registry.endorsersByOrg(channel, chaincodeID)[gs.registry.localEndorser.mspid]
+		if !ok {
+			// No local org endorsers for this channel/chaincode. If transient data is involved, return error
+			if hasTransientData {
+				return nil, status.Error(codes.FailedPrecondition, "no endorsers found in the gateway's organization; retry specifying endorsing organization(s) to protect transient data")
+			}
+			// Otherwise, just let discovery pick one.
+			endorsers, err = gs.registry.endorsers(channel, defaultInterest, "")
+			if err != nil {
+				return nil, status.Errorf(codes.Unavailable, "%s", err)
+			}
+			firstEndorser = endorsers[0]
+		} else {
+			firstEndorser = es[0].endorser
+		}
+
+		// 2. Process the proposal on this endorser
+		firstResponse, err := firstEndorser.client.ProcessProposal(ctx, signedProposal)
+		if err != nil {
+			return nil, rpcError(codes.Aborted, "failed to endorse transaction", endpointError(firstEndorser, err))
+		}
+		if firstResponse.Response.Status < 200 || firstResponse.Response.Status >= 400 {
+			return nil, rpcError(codes.Aborted, "failed to endorse transaction", endpointError(firstEndorser, fmt.Errorf("error %d, %s", firstResponse.Response.Status, firstResponse.Response.Message)))
+		}
+
+		// 3. Extract ChaincodeInterest and SBE policies
+		// The chaincode interest could be nil for legacy peers and for chaincode functions that don't produce a read-write set
+		interest := firstResponse.Interest
+		if len(interest.GetChaincodes()) == 0 {
+			interest = defaultInterest
+		}
+		// The preferred discovery layout will contain the firstEndorser's Org.
+		endorsers, err = gs.registry.endorsers(channel, interest, firstEndorser.mspid)
+		if err != nil {
+			return nil, status.Errorf(codes.Unavailable, "%s", err)
+		}
+
+		// 4. Remove the gateway org's endorser, since we've already done that
+		for i, e := range endorsers {
+			if e.mspid == firstEndorser.mspid {
+				endorsers = append(endorsers[:i], endorsers[i+1:]...)
+				responses = append(responses, firstResponse)
+				break
+			}
+		}
+
+		// 5. If transient data is involved, and other (non-local) orgs are required to endorse, return error for privacy reasons
+		if hasTransientData && len(endorsers) > 0 {
+			return nil, status.Error(codes.FailedPrecondition, "non-local organizations are required to endorse; retry specifying endorsing organization(s) to protect transient data")
+		}
```

Comment:
Not sure if this caution is need here at this stage. Won't this make this whole approach pretty much unusable with private data as in most of the scenario private data transactions will involve transient data and would require multi-org endorsements.

May be better to set the artificial pvt-reads set for all collections to narrow the scope?

Reply:
Sure, I've implemented your suggestion